app/controllers/SignInController.scala

```diff
@@ -57,15 +62,29 @@ class SignInController @Inject() (
                 .flatMap { authenticator =>
                   silhouette.env.eventBus.publish(LoginEvent(user, request))
                   silhouette.env.authenticatorService.init(authenticator).flatMap { v =>
-                    silhouette.env.authenticatorService.embed(v, result)}
+                    silhouette.env.authenticatorService.embed(v, Ok(Json.obj(
+                      "result" -> "success"
+                    )))
+                  }
               }
             case None => Future.failed(new IdentityNotFoundException("Couldn't find user"))
           }
         }.recover {
           case _: ProviderException =>
-            Redirect(routes.SignInController.view()).flashing("error" -> Messages("invalid.credentials"))
+            BadRequest(Json.obj(
+              "providers" -> Json.toJson(socialProviderRegistry.providers)(socialProvidersWrite()),
+              "errors" -> Messages("invalid.credentials")
+            ))
         }
       }
     )
   }
+
+  private def socialProvidersWrite()(implicit request: Request[AnyContent]) = new Writes[Seq[SocialProvider]] {
```

Comment:
This feels pretty janky, but we need the `absoluteURL`s of authenticate endpoints, and that method requires an implicit `request` in scope. Other ideas?

Reply:
how about getting it from the configuration? `configuration.underlying.as[String]("silhouette.google.redirectURL")`

Reply:
Word